File: PythonLeetcode/leetcodeM/742_ClosestLeafInBinaryTree.py
"""
742. Closest Leaf in a Binary Tree


Given a binary tree where every node has a unique value, and a target key k, find the value of the nearest leaf node to target k in the tree.

Here, nearest to a leaf means the least number of edges travelled on the binary tree to reach any leaf of the tree. Also, a node is called a leaf if it has no children.

In the following examples, the input tree is represented in flattened form row by row. The actual root tree given will be a TreeNode object.

Example 1:

Input:
root = [1, 3, 2], k = 1
Diagram of binary tree:
          1
         / \
        3   2

Output: 2 (or 3)

Explanation: Either 2 or 3 is the nearest leaf node to the target of 1.
Example 2:

Input:
root = [1], k = 1
Output: 1

Explanation: The nearest leaf node is the root node itself.
Example 3:

Input:
root = [1,2,3,4,null,null,null,5,null,6], k = 2
Diagram of binary tree:
             1
            / \
           2   3
          /
         4
        /
       5
      /
     6

Output: 3
Explanation: The leaf node with value 3 (and not the leaf node with value 6) is nearest to the node with value 2.
Note:
root represents a binary tree with at least 1 node and at most 1000 nodes.
Every node has a unique node.val in range [1, 1000].
There exists some node in the given binary tree for which node.val == k.

"""


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class CloestLeafInBinaryTree:

    """
    Approach #1: Convert to Graph [Accepted]
    Intuition

    Instead of a binary tree, if we converted the tree to a general graph, we could find the shortest path to a leaf using breadth-first search.

    Algorithm

    We use a depth-first search to record in our graph each edge travelled from parent to node.

    After, we use a breadth-first search on nodes that started with a value of k, so that we are visiting nodes in order of their distance to k.
    When the node is a leaf (it has one outgoing edge, where the root has a "ghost" edge to null), it must be the answer.

    Complexity Analysis

    Time Complexity: O(N) where NN is the number of nodes in the given input tree. We visit every node a constant number of times.

    Space Complexity: O(N), the size of the graph.
    """
    def doit_bfs(self, root, k: int) -> int:
        import collections
        graph = collections.defaultdict(list)

        def dfs(node, par = None):
            if node:
                graph[node].append(par)
                graph[par].append(node)
                dfs(node.left, node)
                dfs(node.right, node)

        dfs(root)
        queue = collections.deque(node for node in graph if node and node.val == k)
        seen = set(queue)

        while queue:
            node = queue.popleft()
            if node:
                if len(graph[node]) <= 1:
                    return node.val
                for nei in graph[node]:
                    if nei not in seen:
                        seen.add(nei)
                        queue.append(nei)

    """
    Approach #2: Annotate Closest Leaf [Accepted]
    Intuition and Algorithm
    
    Say from each node, we already knew where the closest leaf in it's subtree is. Using any kind of traversal plus memoization, we can remember this information.
    
    Then the closest leaf to the target (in general, not just subtree) has to have a lowest common ancestor with the target that is on the path from the root to the target. 
    We can find the path from root to target via any kind of traversal, and look at our annotation for each node on this path to determine all leaf candidates, choosing the best one.
    
    Complexity Analysis

        Time and Space Complexity: O(N). The analysis is the same as in Approach #1.
    """
    def doit_dp_1(self, root, k):
        annotation = {}
        def closest_leaf(root):
            if root not in annotation:
                if not root:
                    ans = float('inf'), None
                elif not root.left and not root.right:
                    ans = 0, root
                else:
                    d1, leaf1 = closest_leaf(root.left)
                    d2, leaf2 = closest_leaf(root.right)
                    ans = min(d1, d2) + 1, leaf1 if d1 < d2 else leaf2
                annotation[root] = ans
            return annotation[root]

        #Search for node.val == k
        path = []
        def dfs(node):
            if not node:
                return
            if node.val == k:
                path.append(node)
                return True
            path.append(node)
            ans1 = dfs(node.left)
            if ans1: return True
            ans2 = dfs(node.right)
            if ans2: return True
            path.pop()

        dfs(root)
        dist, leaf = float('inf'), None
        for i, node in enumerate(path):
            d0, leaf0 = closest_leaf(node)
            d0 += len(path) - 1 - i
            if d0 < dist:
                dist = d0
                leaf = leaf0

        return leaf.val


if __name__ == '__main__':

    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)

    CloestLeafInBinaryTree().doit_(root, 1)

File: cppLeetcode/leetcodeM/742_ClosestLeafInBinaryTree.cpp
/*
742. Closest Leaf in a Binary Tree

Given a binary tree where every node has a unique value, and a target key k, find the value of the nearest leaf node to target k in the tree.

Here, nearest to a leaf means the least number of edges travelled on the binary tree to reach any leaf of the tree. Also, a node is called a leaf if it has no children.

In the following examples, the input tree is represented in flattened form row by row. The actual root tree given will be a TreeNode object.

Example 1:

Input:
root = [1, 3, 2], k = 1
Diagram of binary tree:
          1
         / \
        3   2

Output: 2 (or 3)

Explanation: Either 2 or 3 is the nearest leaf node to the target of 1.
Example 2:

Input:
root = [1], k = 1
Output: 1

Explanation: The nearest leaf node is the root node itself.
Example 3:

Input:
root = [1,2,3,4,null,null,null,5,null,6], k = 2
Diagram of binary tree:
             1
            / \
           2   3
          /
         4
        /
       5
      /
     6

Output: 3
Explanation: The leaf node with value 3 (and not the leaf node with value 6) is nearest to the node with value 2.
Note:
root represents a binary tree with at least 1 node and at most 1000 nodes.
Every node has a unique node.val in range [1, 1000].
There exists some node in the given binary tree for which node.val == k.
*/

#include <functional>
#include <vector>
#include <unordered_map>

using std::unordered_map;
using std::vector;
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class ClosestLeaf {


    /*
        742.Closest-Leaf-in-a-Binary-Tree
        一般解决关于二叉树的问题，90%都可以用DFS解决。此题也不例外，但需要仔细考虑设计。

        首先要明确，从节点k到最近叶子节点的路径，有两种可能：第一种只有一种情况，即从k节点本身开始直接往下走之外；第二种就是可能先向上走，然后经过一个拐点，再从另一端的子树往下。所以这个“拐点”就和路径就是一一对应的关系。我们的目的就是考察所有的节点，找出以其为拐点、“从节点k到最近叶子节点”的路径的最短长度，

        对于节点node，我们首先要知道node到节点k之间的距离，这需要一个递归函数 DFS_find_k。如果节点k在node的左子树，那么我们就要考虑沿着node的右子树往下、到达叶子节点的最短路径，这又需要一个递归函数 DFS_find_closest_leaf. 于是 DFS_find_k(node->left)+2+DFS_find_closest_leaf(node->right) 就是整条路径的长度。反之，如果节点k在node的右子树，那么 DFS_find_closest_leaf(node->right)+2+DFS_find_k(node->left) 就是整条路径的长度。注意，两个长度只可能有一个有效。

        从上可知，我们可以提前做一次整棵树的DFS_find_closest_leaf搜索，记录下所有节点对应的 Dist_to_Leaf[node]，顺带把ClosetLeaf[node]也记录下来，因为结果需要输出的是这个叶子节点的val。然后再过一遍DFS_find_k，对于经过的每一个节点，我们都可以算出如上所述的路径长度。然后再在全局变量中取最小。

        注意：慎用INT_MAX，因为在你不知情的情况下很可能会对INT_MAX做了运算，而INT_MAX+1反而成了最小值。
    */
    unordered_map<TreeNode*,int>Dist2Leaf;
    unordered_map<TreeNode*,int>ClosestLeaf;
    int dist=INT_MAX;
    int result;
public:
    int findClosestLeaf(TreeNode* root, int k) 
    {
        int temp;
        temp = DFS_find_closest_leaf(root);
        temp = DFS_find_k(root,k);
        return result;
    }
    
    int DFS_find_closest_leaf(TreeNode* root)
    {
        if (root==NULL) 
            return INT_MAX;
        
        if (root->left==NULL && root->right==NULL)
        {
            Dist2Leaf[root]=0;
            ClosestLeaf[root]=root->val;
            return 0;
        }
        
        int L=DFS_find_closest_leaf(root->left);
        int R=DFS_find_closest_leaf(root->right);
        
        if (L>R)
        {
            Dist2Leaf[root]=R+1;
            ClosestLeaf[root]=ClosestLeaf[root->right];            
        }
        else
        {
            Dist2Leaf[root]=L+1;
            ClosestLeaf[root]=ClosestLeaf[root->left];            
        }        
        
        return Dist2Leaf[root];
    }
    
    int DFS_find_k(TreeNode* root, int k)
    {
        if (root==NULL) 
            return -1;
        
        if (root->val==k)
        {
            if (dist > Dist2Leaf[root])
            {
                dist = Dist2Leaf[root];
                result = ClosestLeaf[root];
            }
            return 0;
        }
        
        int L=DFS_find_k(root->left,k);
        if (L!=-1)
        {
            if (root->right!=NULL && dist > L+2+Dist2Leaf[root->right])
            {
                dist = L+2+Dist2Leaf[root->right];
                result = ClosestLeaf[root->right];                    
            }
            return L+1;
        }
        
        int R=DFS_find_k(root->right,k);
        if (R!=-1)
        {
            if (root->left!=NULL && dist > R+2+Dist2Leaf[root->left])
            {
                dist = L+2+Dist2Leaf[root->left];
                result = ClosestLeaf[root->left];                    
            }
            return R+1;
        }
        
        return -1;
    }
public:

    int doit_(TreeNode* root, int k) {

        vector<int> paths;
        TreeNode *target = NULL;

        std::function<bool(TreeNode*)> findpath = [&](TreeNode* p) -> bool {
            if (!p) return false;

            paths.push_back(p->val);

            if (p->val == k) {
                target = p;
                return true;
            }

            if (findpath(p->left) || findpath(p->right)) return true;

            paths.pop_back();
            return false;
        };

        findpath(root);

        if (!target->left && !target->right) return target->val;
        
        unordered_map<int, int> buf;
        for (int i = paths.size()-1; i >= 0; i--) 
            buf[paths[i]] = paths.size() - i - 1;
        
        int ans = INT_MAX, minlen = INT_MAX;

        std::function<std::pair<int, int>(TreeNode*)> search = [&](TreeNode* p) -> std::pair<int, int> {

            if (!p->left && !p->right) return {1, p->val};

            int ret = INT_MAX, val = p->val;
            if (p->left) {
                auto it = search(p->left);
                if (it.first < ret) {
                    ret = it.first, val = it.second;
                }
                if (buf.count(p->val) && minlen > ret + buf[p->val]) {
                    minlen = ret + buf[p->val];   
                    ans = val;
                }
            }

            if (p->right) {
                auto it = search(p->right);
                if (it.first < ret) {
                    ret = it.first, val = it.second;
                }
                if (buf.count(p->val) && minlen > ret + buf[p->val]) {
                    minlen = ret + buf[p->val];    
                    ans = val;
                }
            }

            return {ret+1, val};
        };    

        search(root);
        return ans;
    }
};